File: binary-search-tree/minimum-absolute-difference-in-bst/minimum-absolute-difference-in-bst.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution
{
    public int GetMinimumDifference(TreeNode root)
    {
        var stack = new Stack<TreeNode>();
        var curr = root;
        int min = int.MaxValue;
        int? prev = null;

        while (curr != null || stack.Count > 0)
        {
            while (curr != null)
            {
                stack.Push(curr);
                curr = curr.left;
            }

            curr = stack.Pop();
            if (prev.HasValue)
                min = Math.Min(min, curr.val - prev.Value);
            prev = curr.val;

            curr = curr.right;
        }

        return min;
    }

    public int GetMinimumDifferenceMorris(TreeNode root)
    {
        int min = int.MaxValue;
        int? prev = null;
        TreeNode curr = root, pre = null;

        while (curr != null)
        {
            if (curr.left == null)
            {
                if (prev.HasValue)
                    min = Math.Min(min, curr.val - prev.Value);
                prev = curr.val;
                curr = curr.right;
            }
            else
            {
                pre = curr.left;
                while (pre.right != null && pre.right != curr) pre = pre.right;

                if (pre.right == null)
                {
                    pre.right = curr;
                    curr = curr.left;
                }
                else
                {
                    pre.right = null;
                    if (prev.HasValue)
                        min = Math.Min(min, curr.val - prev.Value);
                    prev = curr.val;
                    curr = curr.right;
                }
            }
        }
        return min;
    }
}

File: binary-search-tree/minimum-absolute-difference-in-bst/minimum-absolute-difference-in-bst.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func getMinimumDifference(root *TreeNode) int {
    stack := []*TreeNode{}
    curr := root
    var prev *int
    min := math.MaxInt

    for curr != nil || len(stack) > 0 {
        for curr != nil {
            stack = append(stack, curr)
            curr = curr.Left
        }
        n := len(stack) - 1
        curr = stack[n]
        stack = stack[:n]

        if prev != nil {
            if d := curr.Val - *prev; d < min {
                min = d
            }
        }
        v := curr.Val
        prev = &v

        curr = curr.Right
    }

    return min
}

func getMinimumDifferenceMorris(root *TreeNode) int {
	curr := root
	var prev *int
	min := math.MaxInt

	for curr != nil {
		if curr.Left == nil {
			if prev != nil {
				if d := curr.Val - *prev; d < min {
					min = d
				}
			}
			v := curr.Val
			prev = &v
			curr = curr.Right
		} else {
			pre := curr.Left
			for pre.Right != nil && pre.Right != curr {
				pre = pre.Right
			}
			if pre.Right == nil {
				pre.Right = curr
				curr = curr.Left
			} else {
				pre.Right = nil
				if prev != nil {
					if d := curr.Val - *prev; d < min {
						min = d
					}
				}
				v := curr.Val
				prev = &v
				curr = curr.Right
			}
		}
	}
	return min
}